src/protocol/sort/multi_bit_permutation.rs

```diff
@@ -0,0 +1,256 @@
+            .1
+            .clone();
+        sop_inputs_transposed.push(out);
+    }
+
+    // Take sum of products of output of equality check and accumulated sum
+    let mut permutation_futures = Vec::with_capacity(num_records);
+    for rec in 0..num_records {
+        let ctx_sop = ctx.narrow(&Sop);
+        let mut sop_inputs = Vec::with_capacity(num_possible_bit_values);
+        for sop_input_transposed in sop_inputs_transposed.iter().take(num_possible_bit_values) {
+            sop_inputs.push((&sop_input_transposed[rec].0, &sop_input_transposed[rec].1));
+        }
+        permutation_futures.push(async move {
+            ctx_sop
+                .sum_of_products(RecordId::from(rec), sop_inputs.as_slice())
+                .await
+        });
+    }
+    try_join_all(permutation_futures).await
+}
+
+/// For a given `idx` check if each of the record has same value as idx.
+/// Steps
+/// 1. Get bit representation of `idx`
+/// 2. Calculate equality check
+///   i. keep record bit if `idx` bit is 1
+///   ii.toggle record bit if idx bit is 0 (Done by taking `share_of_one` - value)
+/// 3. Multiply equality checks for all bits per record - in clear per record this will be 1 only for 1 index while 0 for all others
+async fn get_bit_equality_checkers<F, C, S>(
+    idx: usize,
+    input: &[Vec<S>],
+    ctx: C,
+) -> Result<Vec<S>, Error>
+where
+    F: Field,
+    C: Context<F, Share = S>,
+    S: SecretSharing<F>,
+{
+    let num_multi_bits = input.len();
+    let num_records = input[0].len();
+
+    let mut equality_check_futures = Vec::with_capacity(num_records);
+
+    let j_bit_representation = get_binary_from_int(idx, num_multi_bits);
+
+    for rec in 0..num_records {
+        let ctx_across_bits = ctx.narrow(&MultiplyAcrossBits);
+        let share_of_one = ctx_across_bits.share_of_one();
+
+        let mut bits_for_record = Vec::with_capacity(num_multi_bits);
+        for jth_bits in input.iter().take(num_multi_bits) {
+            bits_for_record.push(jth_bits[rec].clone());
+        }
+        let mut mult_input = Vec::with_capacity(num_multi_bits);
+        for jth_bit in j_bit_representation.iter().take(num_multi_bits) {
+            if *jth_bit {
+                mult_input.push(bits_for_record.remove(0));
+            } else {
+                mult_input.push(-bits_for_record.remove(0) + &share_of_one);
+            }
+        }
+        equality_check_futures.push(async move {
+            // multiply all mult_input for this j for each record => f(j)
+            multiply_all_shares(
+                ctx_across_bits,
+                RecordId::from(idx * num_records + rec),
+                mult_input.as_slice(),
+            )
+            .await
+        });
+    }
+    try_join_all(equality_check_futures).await
+}
+
+/// Get binary representation of an integer as a vector of bool
+fn get_binary_from_int(input_num: usize, num_bits: usize) -> Vec<bool> {
```

Comment:
I don't think you need a vector here. or an array. you just need to loop over all bits of an integer, returning `impl Iterator<Item = bool>` is enough. The good news is probably you don't even need to manually implement it, because `bitvec` crate has it. So it should be something similar to 

```rust
   let bit_iter = idx.view_bits::<Lsb0>().iter().take(..) // not sure if usize is easy to deal with here but we have a limit of 1B records so downcasting idx to u32 should be fine
```


in general, try to avoid using vectors unless it is absolutely necessary. Vectors are relatively expensive because they need to heap allocate in order to keep the elements. If you know that the number of items you need to keep is small, prefer tinyvec arrays.

Reply:
to make it a bit more clear - you can get rid of this function and replace it with bitvec `BitView` `view_bits` method.

Reply:
Thanks for the suggestion! This is much better.